core/drivers/crypto/stm32/stm32_hash.c

```diff
@@ -0,0 +1,904 @@
+		c->block_size = SHA256_BLOCK_SIZE;
+		c->save_mode = SAVE_SMALL;
+		break;
+	case STM32_HASH_SHA384:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA2_384))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA384_DIGEST_U32;
+		c->block_size = SHA384_BLOCK_SIZE;
+		c->save_mode = SAVE_BIG;
+		break;
+	case STM32_HASH_SHA512:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA2_512))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA512_DIGEST_U32;
+		c->block_size = SHA512_BLOCK_SIZE;
+		c->save_mode = SAVE_BIG;
+		break;
+	case STM32_HASH_SHA3_224:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA3))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA3_224_DIGEST_U32;
+		c->block_size = SHA3_224_BLOCK_SIZE;
+		c->save_mode = SAVE_SHA3;
+		break;
+	case STM32_HASH_SHA3_256:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA3))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA3_256_DIGEST_U32;
+		c->block_size = SHA3_256_BLOCK_SIZE;
+		c->save_mode = SAVE_SHA3;
+		break;
+	case STM32_HASH_SHA3_384:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA3))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA3_384_DIGEST_U32;
+		c->block_size = SHA3_384_BLOCK_SIZE;
+		c->save_mode = SAVE_SHA3;
+		break;
+	case STM32_HASH_SHA3_512:
+		if (!(c->dev->pdata.compat->caps & CAPS_SHA3))
+			return TEE_ERROR_NOT_IMPLEMENTED;
+
+		c->digest_u32 = SHA3_512_DIGEST_U32;
+		c->block_size = SHA3_512_BLOCK_SIZE;
+		c->save_mode = SAVE_SHA3;
+		break;
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	/*
+	 * The queue size is block_size + one register at first
+	 * then block_size.
+	 * So we may need to save at max queue_size + 3 bytes.
+	 * Let allocate a number of uin32_t: queue_size + 4.
+	 */
+	c->remain.buf = calloc(c->block_size + sizeof(uint32_t), 1);
+	if (!c->remain.buf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	get_save_registers(c, &nb_reg, &first, &hmac_nb_reg, &hmac_first);
+
+	c->csr = calloc(nb_reg + hmac_nb_reg, sizeof(uint32_t));
+	if (!c->csr) {
+		free(c->remain.buf);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	return TEE_SUCCESS;
+}
+
+void stm32_hash_free(struct stm32_hash_context *c)
+{
+	if (!c)
+		return;
+
+	free(c->remain.buf);
+	free(c->csr);
+}
+
+TEE_Result stm32_hash_update(struct stm32_hash_context *c,
+			     const uint8_t *buffer, size_t len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t next_queue_size = c->queue_size;
+	vaddr_t base = 0;
+
+	assert(c);
+
+	base = c->dev->pdata.base;
+
+	if (!len || !buffer)
+		return TEE_SUCCESS;
+
+	mutex_lock(&c->dev->lock);
+	if (clk_enable(c->dev->pdata.clock)) {
+		EMSG("Fail to enable clk %s",
+		     clk_get_name(c->dev->pdata.clock));
+		panic();
+	}
+
+	res = restore_context(c);
+	if (res)
+		goto exit;
+
+	/* We cannot fill the fifo */
+	if (c->remain.len + len < c->queue_size) {
+		if (!c->remain.buf) {
+			res = TEE_ERROR_BAD_STATE;
+			goto exit;
+		}
+
+		memcpy(((uint8_t *)c->remain.buf) + c->remain.len, buffer,
+		       len);
```

Comment:
nitpicking: fits in a single line:
`memcpy((uint8_t *)c->remain.buf + c->remain.len, buffer, len);`
Ditto at line 637: `memcpy((uint8_t *)c->remain.buf + c->remain.len, buffer, align);`